crates/torii/core/src/processors/upgrade_event.rs

```diff
@@ -0,0 +1,120 @@
+use anyhow::{Error, Ok, Result};
+use async_trait::async_trait;
+use dojo_types::schema::Ty;
+use dojo_world::contracts::abigen::world::Event as WorldEvent;
+use dojo_world::contracts::model::ModelReader;
+use dojo_world::contracts::world::WorldContractReader;
+use starknet::core::types::Event;
+use starknet::providers::Provider;
+use tracing::{debug, info};
+
+use super::EventProcessor;
+use crate::sql::Sql;
+
+pub(crate) const LOG_TARGET: &str = "torii_core::processors::upgrade_event";
+
+#[derive(Default, Debug)]
+pub struct UpgradeEventProcessor;
+
+#[async_trait]
+impl<P> EventProcessor<P> for UpgradeEventProcessor
+where
+    P: Provider + Send + Sync + std::fmt::Debug,
+{
+    fn event_key(&self) -> String {
+        "EventUpgraded".to_string()
+    }
+
+    // We might not need this anymore, since we don't have fallback and all world events must
+    // be handled.
+    fn validate(&self, _event: &Event) -> bool {
+        true
+    }
+
+    async fn process(
+        &self,
+        world: &WorldContractReader<P>,
+        db: &mut Sql,
+        _block_number: u64,
+        block_timestamp: u64,
+        _event_id: &str,
+        event: &Event,
+    ) -> Result<(), Error> {
+        // Torii version is coupled to the world version, so we can expect the event to be well
+        // formed.
+        let event = match WorldEvent::try_from(event).unwrap_or_else(|_| {
+            panic!(
+                "Expected {} event to be well formed.",
+                <UpgradeEventProcessor as EventProcessor<P>>::event_key(self)
+            )
+        }) {
+            WorldEvent::EventUpgraded(e) => e,
+            _ => {
+                unreachable!()
+            }
+        };
+
+        // Called model here by language, but it's an event. Torii rework will make clear
+        // distinction.
+        let model = db.model(event.selector).await?;
+        let name = model.name;
+        let namespace = model.namespace;
+        let mut prev_schema = model.schema;
+        match &mut prev_schema {
+            Ty::Struct(s) => {
+                s.name = name.clone();
+            }
+            _ => unreachable!(),
+        }
+
+        let model = world.model_reader(&namespace, &name).await?;
+        let new_schema = model.schema().await?;
+        let schema_diff = new_schema.diff(&prev_schema);
+        // No changes to the schema. This can happen if torii is re-run with a fresh database.
+        // As the register model fetches the latest schema from the chain.
```

Comment:
We might rework the `model.schema()` to accept a block id, which could perhaps be more accurate at some point when replaying history.
But not a blocker for this V1.